### PR TITLE
Implement AWS JSON protocol 1.0

### PR DESCRIPTION
**Pull request by katlasik**
Implement AMZ Json Protocol 1.0.

- [x] Detect AMZ Json Protocol with content type header and parse the request
- [x] Respond with correct error format
~~- [ ] (optional) Customize serialization (use camelCase in case classes and respond with PascalCase)~~
- [x] Support create queue
- [x] Support list queues
- [x] Support send message (need more info regarding MD5OfMessageSystemAttributes)
- [x] Support send message batch
- [x] Support receive message
- [x] Support delete message
- [x] Support delete message batch
- [x] Support get queue URL
- [x] Support purge queue
- [x] Support change message visibility 
- [x] Support change message visibility batch
- [x] Support delete queue
- [x] Support get queue attributes
- [x] Support set queue attributes
- [x] Support add permission
- [x] Support tag queue
- [x] Support untag queue
- [x] Support list tags of queue
- [x] Check if Redrive policy is working (it needs set queue attributes)
~~- [ ] Add ListDeadLetterSourceQueues Action `https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ListDeadLetterSourceQueues.html`~~
~~- [ ] Add RemovePermission Action `https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_RemovePermission.html`~~
- [x] Unify how message attributes are parsed and validated

### CODE REVIEW

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala [hunk 1]
```diff
@@ -80,15 +82,19 @@ trait CreateQueueDirectives {
...
-              respondWith {
-                <CreateQueueResponse>
-                  <CreateQueueResult>
-                    <QueueUrl>{url}</QueueUrl>
-                  </CreateQueueResult>
-                  <ResponseMetadata>
-                    <RequestId>{EmptyRequestId}</RequestId>
-                  </ResponseMetadata>
-                </CreateQueueResponse>
+              protocol match {
```
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala:85]: that's a bigger refactoring but maybe ... we could also have a serializer for `CreateQueueResponse` into XML, and then we could keep the logic of creating the `CreateQueueResponse` instance here, and just call a function with the response + `protocol` + the two implicit serializers (for json and xml)

The code could be a bit cleaner I guess ... but maybe the implicit xml serializers are too much overhead

Comment (katlasik) [reply rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala:85]: we plan to do that, but we want to make it work first

Comment (adamw) [reply rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/CreateQueueDirectives.scala:85]: Right, that's probably a good idea :)

**Comment by flsh86** [comment]
Refactoring with AkkaSupport trait

- [] Respond with correct error format
- [x] Support create queue
- [] Support list queues
- [] Support send message (need more info regarding MD5OfMessageSystemAttributes)
- [x] Support send message batch
- [] Support receive message
- [] Support delete message
- [] Support delete message batch
- [] Support get queue URL
- [] Support purge queue
- [] Support change message visibility 
- [] Support change message visibility batch
- [] Support delete queue
- [] Support get queue attributes
- [] Support set queue attributes
- [] Support add permission
- [] Support tag queue
- [] Support untag queue
- [] Support list tags of queue
- [] Check if Redrive policy is working (it needs set queue attributes)

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/AkkaSupport.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import scala.xml.{Null, UnprefixedAttribute}
+
+trait AkkaSupport {
+  _: RespondDirectives =>
+
+    private def namespace[T](body: UnprefixedAttribute => Marshaller[T, RequestEntity])(implicit v: XmlNsVersion): Marshaller[T, RequestEntity] = {
+      body(new UnprefixedAttribute("xmlns", "http://queue.amazonaws.com/doc/%s/".format(v.version), Null))
+    }
+
+    implicit def elasticMQMarshaller[T](implicit xmlSerializer: XmlSerializer[T], json: RootJsonFormat[T], protocol: AWSProtocol, version: XmlNsVersion) =
```
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/AkkaSupport.scala:22]: it's good to add type annotations to public members

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/ChangeMessageVisibilityBatchDirectives.scala [hunk 3]
```diff
@@ -6,45 +6,80 @@ import org.elasticmq.msg.UpdateVisibilityTimeout
...
-                <ChangeMessageVisibilityBatchResultEntry>
-                  <Id>{id}</Id>
-                </ChangeMessageVisibilityBatchResultEntry>
+          batchRequest(batch.Entries) { (messageData, id, _) =>
+            val result = queueActor ? UpdateVisibilityTimeout(
+              DeliveryReceipt(messageData.ReceiptHandle),
+              MillisVisibilityTimeout.fromSeconds(messageData.VisibilityTimeout)
+            )
+
+            result.flatMap{
```
Comment (katlasik) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/ChangeMessageVisibilityBatchDirectives.scala:29]: We can change it back to `map`.

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala [hunk 4]
```diff
@@ -4,52 +4,94 @@ import akka.http.scaladsl.server.Route
...
-          <GetQueueAttributesResponse>
-            <GetQueueAttributesResult>
-              {attributesToXmlConverter.convert(attributes)}
-            </GetQueueAttributesResult>
-            <ResponseMetadata>
-              <RequestId>{EmptyRequestId}</RequestId>
-            </ResponseMetadata>
-          </GetQueueAttributesResponse>
-        }
+      val requestParams = p.as[GetQueueAttributesActionRequest]
```
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala:19]: 👍 this reads much better :)

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala [hunk 5]
```diff
@@ -4,52 +4,94 @@ import akka.http.scaladsl.server.Route
...
 
 import scala.concurrent.Future
+import scala.xml.Elem
 
 trait QueueAttributesDirectives {
-  this: ElasticMQDirectives with QueueAttributesOps =>
+  this: ElasticMQDirectives with QueueAttributesOps with AkkaSupport =>
 
-  def getQueueAttributes(p: AnyParams): Route = {
+  def getQueueAttributes(p: RequestPayload)(implicit protocol: AWSProtocol, xmlNsVersion: XmlNsVersion): Route = {
```
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala:17]: why is the `xmlNsVersion` an implicit here? shouldn't it rather be a parameter for the xml request payload?

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/XmlSerializer.scala [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package org.elasticmq.rest.sqs
+
+import scala.xml.Elem
+
+trait XmlSerializer[T] {
```
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/XmlSerializer.scala:5]: maybe these should be part of a `marshalling` package?

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/directives/AnyParamDirectives.scala [hunk 7]
```diff
@@ -3,31 +3,94 @@ package org.elasticmq.rest.sqs.directives
...
-      .toMap
+      .map(_.value())
   }
 
-  def anyParamsMap(body: Map[String, String] => Route) = {
-    parameterMap { queryParameters =>
-      extractRequest { request =>
-        entityOrEmpty { fd =>
-          body((fd.fields.toMap ++ queryParameters ++ extractAwsXRayTracingHeader(request)).filter(_._1 != ""))
+  private def extractQueueNameAndUrlFromRequest(body: Map[String, String] => Route): Route = {
```
Comment (adamw) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/directives/AnyParamDirectives.scala:38]: [minor] this looks like a `Directive` ;)

**Comment by adamw** [comment]
I left some comments, looks good overall :)

What about testing? Will it be possible to test both the XML and JSON interfaces?

And also ... @micossow maybe you could take a look as well? You're probably most up-to-date with EMQ :)

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala (hunk 5 above)
Comment (katlasik) [reply rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala:17]: It's used to derive other implicit typeclass: `elasticMQMarshaller`.  Maybe I could merge protocol and version into single case class and pass it as implicit?

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/directives/AWSProtocolDirectives.scala [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+import akka.http.scaladsl.model.{ContentType, MediaType}
+import akka.http.scaladsl.server.Directives.extractRequest
+import org.elasticmq.rest.sqs.AWSProtocol
+
+trait AWSProtocolDirectives {
+
+  import AWSProtocolDirectives._
+  def extractProtocol = extractRequest.map { request =>
+    request.entity.contentType match {
```
Comment (katlasik) [inline rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/directives/AWSProtocolDirectives.scala:11]: I'm not sure about this. Should we already support another version of protocol? I know there's already 1.1, but I'm not sure what is the difference.

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/directives/AnyParamDirectives.scala (hunk 7 above)
Comment (katlasik) [reply rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/directives/AnyParamDirectives.scala:38]: Do you mean I should move it somewhere?

Comment (adamw) [reply rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/directives/AnyParamDirectives.scala:38]: no, the method should be a directive. Sth like `extractQueueNameAndUrlFromRequest(Map): Directive0`, no?

File: rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala (hunk 5 above)
Comment (adamw) [reply rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala:17]: yes that would make sense, as the `xmlNsVersion` only applies to the xml marshaller probably?

Comment (katlasik) [reply rest/rest-sqs/src/main/scala/org/elasticmq/rest/sqs/QueueAttributesDirectives.scala:17]: yes

**Review by micossow — APPROVED** [review]
Great job 👏

**Review by micossow — APPROVED** [review]

**Pull request merged**